qiskit/backends/ibmq/ibmqjob.py

```diff
@@ -359,86 +312,91 @@ def _wait_for_job(self, timeout=60, wait=5):
+                continue
+
+            if api_result['status'] == 'CANCELLED':
+                self._status = JobStatus.CANCELLED
+                job_result = JobResult(self._job_id, 'CANCELLED', api_result['status'])
+                logger.info('The job was cancelled!')
+                break
+
+            # After this point the job is completed!
+            logger.info('Job is complete! status = %s', api_result['status'])
+            job_result_list = []
+            for circuit_result in api_result['qasms']:
+                this_result = {'data': circuit_result['data'],
+                               'name': circuit_result.get('name'),
+                               'compiled_circuit_qasm': circuit_result.get('qasm'),
+                               'status': circuit_result['status']}
+                if 'metadata' in circuit_result:
+                    this_result['metadata'] = circuit_result['metadata']
+                job_result_list.append(this_result)
+
+            logger.info('Got a result from remote backend %s with job id: %s',
+                        self.backend_name, self._job_id)
+            job_result = JobResult(self._job_id, api_result['status'],
+                                   job_result_list,
+                                   api_result.get('usedCredits'),
+                                   backend_name=self._backend_name)
+            self._status = JobStatus.DONE
+            break
 
+        return Result(job_result.as_dict())
 
-class IBMQJobError(QISKitError):
-    """class for IBM Q Job errors"""
-    pass
+    def _is_commercial(self):
+        config = self._api.config
+        # this check may give false positives so should probably be improved
+        return config.get('hub') and config.get('group') and config.get('project')
+
+    def _wait_for_submitting(self):
+        if self._job_id is None:
+            if self._future is None:
+                raise IBMQJobError("You have to submit before asking for status or results!")
+            submit_info = self._future.result(timeout=60)
+            if 'error' in submit_info:
+                raise IBMQJobError(str(submit_info['error']))
+
+            self._creation_date = submit_info.get('creationDate')
+            self._status = JobStatus.QUEUED
+            self._job_id = submit_info.get('id')
 
 
 def _reorder_bits(result):
     """temporary fix for ibmq backends.
     for every ran circuit, get reordering information from qobj
     and apply reordering on result"""
     for circuit_result in result._result['result']:
-        if 'metadata' in circuit_result:
-            circ = circuit_result['metadata'].get('compiled_circuit')
-        else:
-            logger.warning('result object missing metadata for reordering'
-                           ' bits: bits may be out of order')
-            return
+        if 'metadata' not in circuit_result:
+            raise QISKitError('result object missing metadata for reordering bits')
```

Comment:
I originally had exception but Ali pointed out that this would mean that people couldn't get jobs from before there was a 'metadata' field.

Reply:
Yes, the metadata field is very recent, and adding this exception means not being able to get any jobs from before say 2-3 weeks ago. I think we should just return the raw result, even if we miss some reorderings (after all that's how the user saw those results originally too)

Reply:
Makes sense.